### PR TITLE
Get map layers with or without geodata

### PR DESCRIPTION
**Pull request by srezacova**
## Description

Refactor getMapLayers function to return map layers with or without geoData. 

## How has this been tested?

1. Create initial-data-load/data/testCountry4 folder
2. Copy  initial-data-load/data/Sample/MapLayers.yml to testCountry4 folder
3. Remove geoReferenceId from layers config
4. Add COUNTRY=testCountry4 to .env file
5. Run ./start.sh
6. Run ./runinitialload.sh
7. In UI map layers click on Select layers
8. You can see layers there, click on some layer
8. This will cause TypeError. This should never happen because this functionality is not supposed for mapLayers we have there now. These mapLayers should have geoReferenceId always.
9. Check the same for Sample, to see if layers with geoData are uploaded properly

## Types of changes

- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [X] I have read the **CONTRIBUTING** document.
- [X] I have checked to ensure that there aren't other open [Pull Requests](../../../pulls) for the same update/change.
- [X] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [X] All new and existing tests passed.

### CODE REVIEW

File: api/src/actions/mapLayersActions.js [hunk 1]
```diff
@@ -1,25 +1,16 @@
 const model = require('../models/mapLayerModel');
 const { getLayerGeoDataWithUrl } = require('../models/layerGeoDataModel');
 
-const getMapLayersWithGeoData = async () => {
+const getMapLayers = async () => {
```
Comment (bariela) [inline api/src/actions/mapLayersActions.js:4]: can you add doc comment to the function when you are modifying it?

Comment (bariela) [reply api/src/actions/mapLayersActions.js:4]: this one function actually still fulfils the previous name, it is returning the mapLayers with link to geojson file from different collection. We should keep the name after all.

File: api/src/tests/staticLayers.js [hunk 2]
```diff
@@ -13,13 +13,6 @@ jest.mock('../config/config.js', () => {
 });
 
 describe('GET /api/staticLayers', () => {
-  it('should return empty array, if no geojson data in db', async () => {
```
Comment (bariela) [inline api/src/tests/staticLayers.js:16]: please don't remove test, but modify it. In the PR you wrote quite nice scenario and this one test needs only change in description and on L21 to expect the one layer

**Review by bariela — CHANGES_REQUESTED** [review]
scenarios need to be tested in automated way to avoid future issues :)

**Comment by srezacova** [comment]
Hi, I have just few questions. 
1) getGeojsonData function in vectorSourceLoader and boxVectorSourceLoader return exactly the same structure? I just have some problem with debugging on frontend so I cannot show it in console. In case of boxVectorSourceLoader it is still geojsonData?
2)Error I mentioned in point 9 is:
```
Unhandled Rejection (TypeError): Cannot read properties of undefined (reading 'protocol')

  92 | export const getGeojsonData = async (url) => {
> 93 |   const response = await axios.get(url);
  94 |   return response.data;
  95 | };
```
I can get rid of it by adding something like this:
```
      if (url) {
        response = await getGeojsonData(url);
      }
```
but I am not sure if it is necessary. Then I need to also put
`        handleIsLoading(title, 'remove');`
to  LN76. And in points.js on LN11 there is also condition that depends on geoReferenceId, that I must change if I do these changes.
3)In the first test when I will try to expect something like this:
`    expect(res.body[0].layerType).toEqual(mapLayerWithoutGeodata.layerType);`
It will not work, it uses MapLayer model which uses MapLayerSchema which do not include layerType. But when we are getting data with getMapLayers function, we use exactly the same MapLayer model and it return all properties we have in database layerType included. Whats the reason? Should I use SingleMapLayer discriminator instead?

File: api/src/tests/staticLayers.js [hunk 3]
```diff
@@ -13,13 +13,15 @@ jest.mock('../config/config.js', () => {
...
+  it('should return layer, even if no geojson data are in db', async () => {
     await MapLayer.create(mapLayersInDb[0]);
 
     const res = await request(app).get('/api/staticLayers/');
     expect(res.status).toEqual(200);
-    expect(res.body).toHaveLength(0);
+    expect(res.body).toHaveLength(1);
   });
+});
```
[the remark this answers is not in the document]
Comment (srezacova) [reply api/src/tests/staticLayers.js]: Removed, thanks for noting.

File: api/src/tests/staticLayers.js [hunk 4]
```diff
@@ -13,13 +13,15 @@ jest.mock('../config/config.js', () => {
...
 
 describe('GET /api/staticLayers', () => {
-  it('should return empty array, if no geojson data in db', async () => {
+  it('should return layer, even if no geojson data are in db', async () => {
     await MapLayer.create(mapLayersInDb[0]);
 
     const res = await request(app).get('/api/staticLayers/');
     expect(res.status).toEqual(200);
-    expect(res.body).toHaveLength(0);
+    expect(res.body).toHaveLength(1);
```
[the remark this answers is not in the document]
Comment (srezacova) [reply api/src/tests/staticLayers.js:21]: I added some. Please check again.

[the remark this answers is not in the document]
Comment (bariela) [reply api/src/tests/staticLayers.js:21]: also that there is really no geoReferenceId or geoDataUrl

**Comment by bariela** [comment]
@srezacova hi,
1. it is literally the same function, so definitely yes. Difference might be only in code of the respective loaders
2. sounds reasonable, but also with else (returning probably just [] ?)
3. definitely the discriminator, the model itself is just some common parts for both of them

**Review by bariela — CHANGES_REQUESTED** [review]
the discriminator should resolve problem in test

**Comment by srezacova** [comment]
> @srezacova hi,
> 
> 1. it is literally the same function, so definitely yes. Difference might be only in code of the respective loaders
> 2. sounds reasonable, but also with else (returning probably just [] ?)
> 3. definitely the discriminator, the model itself is just some common parts for both of them

2. I fixed this in VectorSourceLoader. You can test it with layer with no geoReferenceId.In case that normal mapLayers wouldn't have in some case geoReferenceId, what to do with condition on LN10 in points.js? Now it uses Kobo loader for these layers. Will this case ever happen for normal mapLayers? If not I will leave it like this.
3. I put discriminator in first test, but it will not work for second test with:
```
        color: {
          type: 'colormap',
          value: 'green',
        },
```
in legend, because SingleMapLayer is expecting color as string. Thats probably the reason why we used MapLayer model there? 
getMapLayers function in mapLayersModel.js return are all map layers properties even those that are not in MapLayer model. (for example layerType) Why? It should only return those properties that are in MapLayerSchema or not?
Thanks for answer

File: api/src/tests/staticLayers.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (srezacova) [reply api/src/tests/staticLayers.js:21]: I fixed that. Please check again.

**Comment by bariela** [comment]
> > @srezacova hi,
> > 
> > 1. it is literally the same function, so definitely yes. Difference might be only in code of the respective loaders
> > 2. sounds reasonable, but also with else (returning probably just [] ?)
> > 3. definitely the discriminator, the model itself is just some common parts for both of them
> 
> 1. I fixed this in VectorSourceLoader. You can test it with layer with no geoReferenceId.In case that normal mapLayers wouldn't have in some case geoReferenceId, what to do with condition on LN10 in points.js? Now it uses Kobo loader for these layers. Will this case ever happen for normal mapLayers? If not I will leave it like this.
> 2. I put discriminator in first test, but it will not work for second test with:
> 
> ```
>         color: {
>           type: 'colormap',
>           value: 'green',
>         },
> ```
> 
> in legend, because SingleMapLayer is expecting color as string. Thats probably the reason why we used MapLayer model there? getMapLayers function in mapLayersModel.js return are all map layers properties even those that are not in MapLayer model. (for example layerType) Why? It should only return those properties that are in MapLayerSchema or not? Thanks for answer

Hi,
2. I would resolve already in axiosRequests.js, so the loader gets just correct data format. Maybe even try catch and finally? So error in one data won't break the map
3. color as string is correct for all models, type is on the same level, it is slightly different format from style. It is filtered at creating/saving to db. That is the point of having models.
For getMapLayers we use MapLayer model to get both single and group map layers = layers from all discriminators

I realised just now one new issue - we should not return mapLayers of layerType='regions' with no geoData. Using the old filter & adding condition for layerType 'points' here https://github.com/openkfw/Oscar/pull/77/files#diff-4401c75e5a0122ab94518d56931c9b5eb06c7a1af2764942c5c4b7d050dd9750L18 could work with this requirement.

**Pull request merged**